File: src/model/ScoreMap.go
package model

type Turn = bool

const (
	Player1Turn = true
	Player2Turn = false

	minWinnerScore = int(MaxBox/2 + 1)
)

type ScoreMap struct {
	Turn
	Player1Score int
	Player2Score int
}

func (m *ScoreMap) Reset() {
	m.Turn = Player1Turn
	m.Player1Score = 0
	m.Player2Score = 0
}

func (m *ScoreMap) Add(s int) {
	if s == 0 {
		m.Turn = !m.Turn
		return
	}
	if m.Turn == Player1Turn {
		m.Player1Score += s
	} else {
		m.Player2Score += s
	}
}

func (m *ScoreMap) Score() int { return m.Player1Score - m.Player2Score }

func (m *ScoreMap) NotOver() bool {
	return m.Player1Score < minWinnerScore && m.Player2Score < minWinnerScore
}

File: src/model/ScoreMap.cpp
#include "ScoreMap.h"

ScoreMap::ScoreMap() {
  Reset();
}

void
ScoreMap::Reset() {
  Player1Score = 0;
  Player2Score = 0;
  Turn = Player1Turn;
}

void
ScoreMap::Add(int s) {
  if (s == 0) {
    Turn = !Turn;
    return;
  }
  if (Turn == Player1Turn) {
    Player1Score += s;
  } else {
    Player2Score += s;
  }
}

[[nodiscard]] int
ScoreMap::Score() const {
  return Player1Score - Player2Score;
}

[[nodiscard]] int
ScoreMap::GetScore(int player) const {
  return (player == 0) ? Player1Score : Player2Score;
}

[[nodiscard]] bool
ScoreMap::Gaming() const {
  return Player1Score < MinWinnerScore && Player2Score < MinWinnerScore;
}
